# Assignment 4

The assignment 4 contains 3 questions and the due date is May 9th (Sunday) 23:59PM.

## Question 1 Customer Segmentation

The management team of a large shopping mall would like to understand the types of people who are, or could be, visiting their mall. They have good reasons to believe that there are a few different market segments, and they are considering designing and positioning the shopping mall services better in order to attract mainly a few profitable market segments, or to differentiate their services (e.g. invitations to events, discounts, etc) across market segments.

We aim to describe attitudes of people to shopping in a shopping mall using a small dataset (`Mall_Visits.csv`). To make these decisions, the management team run a market research survey of a few potential customers. In this case this was a small survey to only a few people, where each person answered six attitudinal questions and a question regarding how often they visit the mall, all on a scale 1-7, as well as one question regarding their household income:

| Name        | Name           | Name  |
| ------------- |:-------------:| -----:|
| V1     | Shopping is fun | 1-7 |
| V2      | Shopping is bad for your budget      |  1-7 |
| V3 | I combine shopping with eating out      |    1-7 |
| V4     | I try to get the best buys while shopping | 1-7 |
| V5      | I don’t care about shopping      |  1-7 |
| V6 | You can save lot of money by comparing prices      |    1-7 |
| Income      | The household income of the respondent      |  Dollars |
| Mall.Visits | How often they visit the mall      |    1-7 |


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [3]:
mall_visit = pd.read_csv('Mall_Visits.csv')
mall_visit.head()

ID  V1  V2  V3  V4  V5  V6  Income  Mall.Visits
0   1   6   4   7   3   2   3   60000            3
1   2   2   3   1   4   5   4   30000            1
2   3   7   2   6   4   1   3   70000            3
3   4   4   6   4   5   3   6   30000            7
4   5   1   3   2   2   6   4   60000            1

In [4]:
mall_visit.describe()

ID         V1         V2         V3         V4       V5  \
count  40.000000  40.000000  40.000000  40.000000  40.000000  40.0000   
mean   20.500000   3.850000   4.100000   3.950000   4.100000   3.4500   
std    11.690452   1.874936   1.392286   1.986493   1.498717   1.7387   
min     1.000000   1.000000   2.000000   1.000000   2.000000   1.0000   
25%    10.750000   2.000000   3.000000   2.000000   3.000000   2.0000   
50%    20.500000   4.000000   4.000000   4.000000   4.000000   3.5000   
75%    30.250000   5.250000   5.000000   6.000000   5.250000   4.2500   
max    40.000000   7.000000   7.000000   7.000000   7.000000   7.0000   

              V6       Income  Mall.Visits  
count  40.000000     40.00000     40.00000  
mean    4.350000  46000.00000      3.25000  
std     1.477177  17216.56924      1.94475  
min     2.000000  25000.00000      0.00000  
25%     3.000000  30000.00000      2.00000  
50%     4.000000  42500.00000      3.00000  
75%     5.250000  60000.00000      4.25000  
max     7.000000  80000.00000      7.00000

### Question 1.1 (5 points) 

Perform customer segmentation using k-Means algorithm by setting k=1. What can we infer from customer segmentation with one cluster? Explain this using the obtained centroid of cluster.

### Question 1.2 (10 points)

Now perform customer segmentation by setting the number of k to be 3. Explain the obtained clusters. Check the distribution of each feature across the clusters and see whether these features are statisically significantly different from each other. 

### Question 1.3 (5 points)

Now perform customer segmentation by setting the number of k using the Elbow method and metrics of inertia, respectively. Explain what would be the k you get.

## Question 2 Birds of a feather laugh together

Jester is a research project from the UC Berkeley Laboratory for Automation Science and Engineering. Users are asked to rate the jokes they read and may see more similar jokes. The dataset `jester_dataset.zip`, when unzipped, they are in Excel (.xls) format that contains ratings from 23,500 users who have rated 36 or more jokes. Ratings are real values ranging from -10.00 to +10.00 (the value "99" corresponds to "null" = "not rated"). The text of the jokes is given `jester_dataset_joke_text.zip`. The first column gives the number of jokes rated by that user. The next 100 columns give the ratings for jokes 01 - 100.

### Question 2.1 (10 points) 

Build a user-based collaborative filtering and provide a list of top 5 jokes for the first user. Check the text of jokes to see if the recommendations are sensible. 

### Question 2.2 (10 points) 

Build a item-based collaborative filtering and provide a list of top 5 jokes for the first user. Check the text of jokes to see if the recommendations are sensible.

### Question 2.3 (10 points) 

Now build a recommendation system using matrix factorization (singular value decomposition). Get a list of top 5 jokes for the first user. Explain whether model-based recoomendation approach is better than neighborhood-based approach. Write a function that takes the parameter of user_id and return the text of recommended jokes)

## Question 3 Training a deep neural network (50 points)

Practice training a deep neural network on the CIFAR10 image dataset:


In [33]:
# TensorFlow ≥2.0 is required
import tensorflow as tf
from tensorflow import keras
assert tf.__version__ >= "2.0"


keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

(X_train_full, y_train_full), (X_test, y_test) = keras.datasets.cifar10.load_data()

X_train = X_train_full[5000:]
y_train = y_train_full[5000:]
X_valid = X_train_full[:5000]
y_valid = y_train_full[:5000]

### Question 3.1 (10 points) 

Build a Deep Neural Network (DNN) with 20 hidden layers of 100 neurons each (that’s too many, but it’s the point of this question), using the RELU activation function.

Using Adam optimization, train the network on the CIFAR10 dataset. You can load it with keras.datasets.cifar10.load_ data() (see code below). The dataset is composed of 60,000 32 × 32–pixel color images (50,000 for training, 5000 of which is for validation, 10,000 for testing) with 10 classes using sparse_categorical_crossentropy loss function , so you’ll need a softmax output layer with 10 neurons. 

Train the model for 100 epochs, explain whether the model will experience overfitting. 

### Question 3.2 (10 points)

Try reduce the overfitting with weight regularization. You can try to add a L2-norm penalty term as the kernel_regularizer. Repeat the previous training process and explain whether regularization helps achieve a better result.

### Question 3.3 (10 points)

Now try adding Batch Normalization and compare the learning curves: Is it converging faster than before? Does it produce a better model? How does it affect training speed?

### Question 3.4 (10 points)

Try regularizing the model with dropout. Then, without retraining your model, see if you can achieve better accuracy using different rate of Dropout.

### Question 3.5 (10 points) 

Add Earlystopping to the model and save the model at the early stopping epoch using the callbacks. 

Last, load the model and evaluate on the test set.